backend/pkg/rate/limit_http.go

```
@@ -0,0 +1,204 @@
+	id := identity.FromContext(r.Context())
+	ctx := map[string]any{
+		"Identity": id,
+		// "Request":  r, // not needed, but could be handy
+	}
+	return ctx
+}
+
+func NewHTTPLimiter(eventLimiter EventLimiter, eventTemplate string) (*HTTPLimiter, error) {
+	template, err := template.New("").Parse(eventTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("invalid eventTemplate: %w", err)
+	}
+	return &HTTPLimiter{
+		template:     template.New("").Option("missingkey=zero"),
+		mux:          http.NewServeMux(),
+		templateData: defaultTemplateData,
+		DefaultGroup: &limiterGroup{
+			limiter:       eventLimiter,
+			eventTemplate: template,
+		},
+		limiterGroups: make(map[string]*limiterGroup),
+	}, nil
+}
+
+func (h *HTTPLimiter) WithTemplateDataFunc(f func(*http.Request) any) *HTTPLimiter {
+	h.templateData = f
+	return h
+}
+
+func (h *HTTPLimiter) WithTemplateFuncs(funcs map[string]any) *HTTPLimiter {
+	h.template.Funcs(funcs)
+	return h
+}
+
+func (h *HTTPLimiter) WithRewriteRequests(rewrite bool) *HTTPLimiter {
+	h.rewriteRequests = rewrite
+	return h
+}
+
+func (h *HTTPLimiter) AddRateLimitGroup(limiter EventLimiter, group, eventTemplate string) error {
+	t, err := h.template.Clone()
+	if err == nil {
+		_, err = t.Parse(eventTemplate)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to compile event template: %w", err)
+	}
+	h.limiterGroups[group] = &limiterGroup{
+		limiter:       limiter,
+		eventTemplate: t,
+	}
+	return nil
+}
+
+func (h *HTTPLimiter) MatchHTTPPattern(
+	pattern, groupTemplate string,
+) error {
+	var (
+		t   *template.Template
+		err error
+	)
+	if groupTemplate != "" {
+		// Compile eventTemplate:
+		t, err = h.template.Clone()
+		if err == nil {
+			_, err = t.Parse(groupTemplate)
+		}
+		if err != nil {
+			return fmt.Errorf("error parsing group_template: %w", err)
+		}
+	}
+	limiterHandle := handle{
+		HTTPLimiter:   h,
+		groupTemplate: t,
+	}
+	h.mux.Handle(pattern, limiterHandle)
```

this panics if the patterns conflicts; we are aware of that, right?

---

Yes, that would be a configuration error. I don't want to mess with the underlying implementation and instead let it panic as this would result from a configuration error.